TESTING.md

```diff
@@ -0,0 +1,87 @@
+# Testing Code Climate Engines
+
+The `codeclimate` CLI has a `test` command for automatically running a suite of regression tests that should be provided by the engine maintainer.
+
+The tests described in this document are designed to assert that the results of your engine are what you expect. For that reason, it is designed at a particular level of granularity that allows for detailed assertions to be made about the position, categorization, and content of the issues emitted by your engine.
+
+This document describes how to configure your engine to automatically run tests when `codeclimate test YOUR_ENGINE_NAME` is executed, describes the engine test format, and discusses integrating engine tests into your engine development workflow.
+
+## Configuring Your Engine To Run Tests
+
+Your engine can be configured to run tests by supplying path information in your engine's `engine.json` file:
+
+```
+{
+  "name": "fixme",
+  "description": "Discover lurking FIXMEs, BUGs, TODOs, etc. in your code",
+  "maintainer": {
+    "name": "Michael R. Bernstein",
+    "email": "mrb@codeclimate.com"
+  },
+  "languages": ["*"],
+  "test_paths": [
+    "/usr/src/app/test/integration/"
+  ]
+}
+```
+
+*Note: By convention, the `engine.json` file must be in the root directory of your container.*
+
+When `codeclimate test` is run, the `test_paths` are unpacked, and every file in the directory is run as a test (under the hood, we run `codeclimate analyze` with the `--engine` option and the test file path).
```

Comment:
We are going to want to change this behavior so that, in effect, we `cd` into each test path and run `codeclimate analyze` once per test path. That means that, for example if you have these files:

```
/usr/src/app/test/integration/foo_test.rb
/usr/src/app/test/integration/bar_test.rb
```

We will run `codeclimate analyze` _once_, and validate the results from both files at once. This solves how we will accommodate tests that need to run with custom configuration files, or against a full directory tree. You simply list each test dir individually in `test_paths`. For example:

```
"test_paths": ["/usr/src/app/tests/custom_mass_threshold/"]
```

Then in `/usr/src/app/tests/custom_mass_threshold` you can have both a `.codeclimate.yml` and a `duplication_test.rb` file.